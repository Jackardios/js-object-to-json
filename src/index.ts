const singleLineCommentsRegExp = /\/\/.*/gm;
const multiLineCommentsRegExp = /\/\*(.|\n)*?\*\//gm;
const trailingCommasRegExp = /(,)((?:\s|\n)*?(?:}|]))/gm;
const unquotedPropertyRegExp = /(?<name>[a-zA-Z0-9]+)(\n|\s)*:/gm;
const stringsRegExp = /(?<!\\)(?<quote>['"])(?:\\\k<quote>|.)*?\k<quote>/gm;

function unescapeDoubleQuotesString(string: string) {
  return string.replace(/\\'/gm, "'");
}

function singleQuotesToDoubleQuotes(string: string) {
  const unquoted = string.substring(1, string.length - 1);
  const escaped = unescapeDoubleQuotesString(unquoted).replace(
    /(?<!\\)"/gm,
    '\\"'
  );

  return `"${escaped}"`;
}

function normalizeJSStrings(string: string) {
  return string.replace(stringsRegExp, match => {
    return match[0] === "'"
      ? singleQuotesToDoubleQuotes(match)
      : unescapeDoubleQuotesString(match);
  });
}

function removeJSComments(string: string) {
  return string
    .replace(multiLineCommentsRegExp, '')
    .replace(singleLineCommentsRegExp, '');
}

function removeTrailingCommas(string: string) {
  return string.replace(trailingCommasRegExp, '$2');
}

export function jsObjectAsStringToJSON(jsObjectAsString: string): string {
  const normalized = normalizeJSStrings(
    removeTrailingCommas(removeJSComments(jsObjectAsString))
  ).replace(unquotedPropertyRegExp, '"$1":');

  // validate
  JSON.parse(normalized);

  return normalized;
}
