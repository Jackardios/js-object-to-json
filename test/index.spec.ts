import { jsObjectAsStringToJSON } from '../src';

describe('index', () => {
  describe('jsObjectAsStringToJSON', () => {
    it('should convert empty object to json', () => {
      expect(jsObjectAsStringToJSON('{}')).toEqual('{}');
    });

    it('should convert complex js object to json', () => {
      expect(
        jsObjectAsStringToJSON(`
{
  // some single line comment
  fooBar: 'baz',
  barBaz: 123,
  'bar-baz': false,
  "alex": true,
  "alex[name=\\"double-escaped\\"]": {
    someNested: [
      1,
      2,/*
        // inner comment
      */
      { "maxWidth": '123px', 'max-height': "value" },
    ],
    anotherNested: [1, 2,{}, 'foo', ''],
    /*

    another: [1, 2, 'foo','bar' {}, "hello"]
    'multiline comment'
  */
    fooBar: [1,2, false, true, 'some-"string"', 3,]
  },
  'alex[name=\\'single-escaped\\']': false,//
  "alex[name=\\'single-escaped\\']": [1,2,false,  ],//
  "alex[name='single-unescaped']": [1,2,3,false], // : ['some value'],
  'alex[name=\\"double-escaped\\"]': "some string val'ue",/**/
'alex[name="double-unescap\\'ed"]': "some string val'\\"u\\"'e",
  'alex[name="double-unescaped"]': "some string 'value'",
}
      `)
      ).toMatchSnapshot();
    });

    it('should throw an error when converting invalid js object to json', () => {
      expect(() => jsObjectAsStringToJSON('some invalid json')).toThrow(
        SyntaxError
      );
    });
  });
});
