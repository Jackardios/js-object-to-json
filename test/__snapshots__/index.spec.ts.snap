// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index jsObjectAsStringToJSON should convert complex js object to json 1`] = `
"
{
  
  \\"fooBar\\": \\"baz\\",
  \\"barBaz\\": 123,
  \\"bar-baz\\": false,
  \\"alex\\": true,
  \\"alex[name=\\\\\\"double-escaped\\\\\\"]\\": {
    \\"someNested\\": [
      1,
      2,
      { \\"maxWidth\\": \\"123px\\", \\"max-height\\": \\"value\\" }
    ],
    \\"anotherNested\\": [1, 2,{}, \\"foo\\", \\"\\"],
    
    \\"fooBar\\": [1,2, false, true, \\"some-\\\\\\"string\\\\\\"\\", 3]
  },
  \\"alex[name='single-escaped']\\": false,
  \\"alex[name='single-escaped']\\": [1,2,false  ],
  \\"alex[name='single-unescaped']\\": [1,2,3,false], 
  \\"alex[name=\\\\\\"double-escaped\\\\\\"]\\": \\"some string val'ue\\",
\\"alex[name=\\\\\\"double-unescap'ed\\\\\\"]\\": \\"some string val'\\\\\\"u\\\\\\"'e\\",
  \\"alex[name=\\\\\\"double-unescaped\\\\\\"]\\": \\"some string 'value'\\"
}
      "
`;
